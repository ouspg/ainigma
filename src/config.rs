use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
struct CourseConfiguration {
    course_identifier: CourseIdentifier,
    weeks: Vec<Weeks>,
    tasks: Vec<WeeksTasks>,
    taskbuild: Vec<WeeksTasksBuild>,
    taskoutput: Vec<WeeksTasksOutput>,
}

impl CourseConfiguration {
    pub fn new(
        course_identifier: CourseIdentifier,
        weeks: Vec<Weeks>,
        tasks: Vec<WeeksTasks>,
        taskbuild: Vec<WeeksTasksBuild>,
        taskoutput: Vec<WeeksTasksOutput>,
    ) -> CourseConfiguration {
        CourseConfiguration {
            course_identifier,
            weeks,
            tasks,
            taskbuild,
            taskoutput,
        }
    }
}

#[derive(Deserialize)]
struct CourseIdentifier {
    //TODO:Change to UUID
    identifier: String,
    name: String,
    description: String,
    version: String,
}

impl CourseIdentifier {
    pub fn new(
        identifier: String,
        name: String,
        description: String,
        version: String,
    ) -> CourseIdentifier {
        CourseIdentifier {
            identifier,
            name,
            description,
            version,
        }
    }
}
#[derive(Deserialize)]
struct Weeks {
    number: i32,
    theme: String,
}

impl Weeks {
    pub fn new(number: i32, theme: String) -> Weeks {
        Weeks { number, theme }
    }
}
#[derive(Deserialize)]
struct WeeksTasks {
    id: String,
    name: String,
    description: String,
    points: f32,
    flags: Vec<FlagConfig>,
    subtasks: Vec<SubTask>,
}

impl WeeksTasks {
    pub fn new(
        id: String,
        name: String,
        description: String,
        points: f32,
        flags: Vec<FlagConfig>,
        subtasks: Vec<SubTask>,
    ) -> WeeksTasks {
        WeeksTasks {
            id,
            name,
            description,
            points,
            flags,
            subtasks,
        }
    }
}
#[derive(Deserialize)]
struct FlagConfig {
    flag_type: String,
    id: String,
}

impl FlagConfig {
    pub fn new(flag_type: String, id: String) -> FlagConfig {
        FlagConfig { flag_type, id }
    }
}

#[derive(Deserialize)]
struct SubTask {
    id: String,
    name: String,
    description: String,
    subpoints: f32,
}

impl SubTask {
    pub fn new(id: String, name: String, description: String, subpoints: f32) -> SubTask {
        SubTask {
            id,
            name,
            description,
            subpoints,
        }
    }
}
#[derive(Deserialize)]
struct WeeksTasksBuild {
    directory: String,
    entrypoint: String,
    builder: String,
}

impl WeeksTasksBuild {
    pub fn new(directory: String, entrypoint: String, builder: String) -> WeeksTasksBuild {
        WeeksTasksBuild {
            directory,
            entrypoint,
            builder,
        }
    }
}
#[derive(Deserialize)]
struct WeeksTasksOutput {
    name: String,
    output_type: String,
}

impl WeeksTasksOutput {
    pub fn new(name: String, output_type: String) -> WeeksTasksOutput {
        WeeksTasksOutput { name, output_type }
    }
}
