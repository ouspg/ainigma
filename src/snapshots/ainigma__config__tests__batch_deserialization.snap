---
source: src/config.rs
expression: result
---
ModuleConfiguration {
    identifier: 01920a0e-8c4e-7bbb-80a3-fcae235253ae,
    name: "Something about the cloud security",
    description: "",
    version: "0.0.1",
    categories: [
        Category {
            tasks: [
                Task {
                    id: "task002",
                    name: "How do containers work?",
                    description: "",
                    points: 1.0,
                    stages: [
                        TaskElement {
                            id: Some(
                                "task0021",
                            ),
                            name: Some(
                                "Find the flag :)",
                            ),
                            description: None,
                            weight: Some(
                                1,
                            ),
                            flag: FlagVariant {
                                kind: PureRandom,
                            },
                            batch: Some(
                                BatchConfig {
                                    count: 6,
                                },
                            ),
                        },
                    ],
                    build: BuildConfig {
                        directory: "ansible",
                        builder: Shell(
                            Shell {
                                entrypoint: "entrypoint.sh",
                            },
                        ),
                        output: [
                            BuildOutputFile {
                                kind: Readme(
                                    "readme.txt",
                                ),
                            },
                        ],
                        enabled_modes: NonEmptyBuildModes(
                            [
                                Batch,
                            ],
                        ),
                    },
                },
            ],
            number: 4,
            name: "Container security",
        },
    ],
    flag_config: FlagConfig {
        pure_random: PureRandom {
            length: 32,
        },
        user_derived: UserDerived {
            algorithm: HMAC_SHA3_256,
            secret: "Try to guess the super secret secret!",
        },
        rng_seed: RngSeed {
            secret: "HaHa you don't know it",
        },
    },
    deployment: Deployment {
        build_timeout: 300,
        upload: Upload {
            aws_s3_endpoint: "",
            aws_region: "",
            bucket_name: "",
            use_pre_signed: false,
            link_expiration: 7,
            file_expiration: 31,
        },
    },
}
